Adds git clone method to search for metadata file

logicminds opened the pull request:
* Previously the spec search method was limited to just github.com
    and gitlab.com.  Since there are many more git providers and
    internal sources a git clone option was added to support git remotes from
   various sources. 

* It was originally thought that cloning modules would take FOREVER!, Which it actually did with
  some of the sample tests.   Cloning initial took 15 seconds.  With some special git options I 
  was able to reduce this to 1 second.  The git options used where `--bare --single-branch --depth=1 --branch=ref`

* Previous comments in the code discussed adding the git clone ability so this might be a internal issue or ticket that
  can be closed as a result.

* I added my own testing file and added support for ref, commit, branch, and tag as the other classes didn't support those.

Commit:
Adds git clone method to search for metadata file

  * Previously the spec search method was limited to just github.com
    and gitlab.com.  Since there are many more git providers and
    internal sources a git clone option was added to support git remote

  * Fixes #27 - support non github urls

logicminds commented:
Fixes #27

logicminds commented:
It appears to fail on windows.  Maybe lack of git in path or something?  Any ideas?

logicminds commented:
@glennsarti this would assume that git exists on the command line.  Looking over my code I would need to check for existence of git first, fail if not found.

glennsarti commented on lib/puppetfile-resolver/spec_searchers/git/gclone.rb:
```diff
@@ -0,0 +1,84 @@
+
+          return nil if repo_url.nil?
+          return nil unless valid_http_url?(repo_url)
+          metadata_file = 'metadata.json'
```
Like you mentioned in PR comments already.  Need a guard here to check if `git` is even available.  And memoized too would be best.

glennsarti commented on lib/puppetfile-resolver/spec_searchers/git/gclone.rb:
```diff
@@ -0,0 +1,84 @@
+        # @param silent [Boolean] set to true if you wish to send output to /dev/null, false by default
+        # @return [Array]
+        def self.run_command(cmd, silent: false)
+          out_args = silent ? '2>&1 > /dev/null' : '2>&1'
```
`/dev/null` is no-bueno on Windows.

Personally I prefer to _not_ delegate stdout/stderr to the shell and instead use open3 to create the process. 

I'd also prefer to see the output of the process go to resolver debug log.  So something like

``` ruby
cmd = ['git','clone','--bare','--depth=1','--single-branch']
...
Open3.capture3(*cmd)
```

I've found using open3 avoids any shell quoting issues. Redirection is much simpler and is much more cross-platform friendly.

glennsarti commented on lib/puppetfile-resolver/spec_searchers/git/gclone.rb:
```diff
@@ -0,0 +1,84 @@
+            out, successful = run_command("#{CLONE_CMD} #{branch} #{url} #{dir} #{proxy}", silent: true)
+            err_msg += out
+            raise err_msg unless successful
+            Dir.chdir(dir) do
```
Future todo would be to have a "global" git cache so it would only git clone if it needed to.  But that's not needed right now.

glennsarti commented on lib/puppetfile-resolver/spec_searchers/git/gclone.rb:
```diff
@@ -0,0 +1,84 @@
+        def self.clone_and_read_file(url, ref, file, config)
+          # cloning is a last resort if for some reason we cannot
+          # remotely get via ls-remote
+          Dir.mktmpdir do |dir|
```
Future - Make this working dir configurable.  Deep paths can cause issues on Windows without LFNs.

logicminds replied:
configurable how?  It would be configurable based on the environment variable TMP or TEMP by default, but if we want to configure via a resolver variable that would be different.

glennsarti replied:
yeah, something in the resolver config.  But this is more a future thing.

logicminds replied:
Updated.  The gitconfiguration now has a clone_dir.

glennsarti commented on lib/puppetfile-resolver/spec_searchers/git.rb:
```diff
@@ -14,12 +15,13 @@ def self.find_all(puppetfile_module, dependency, cache, resolver_ui, config)
-
-        # TODO: Once we've exhausted using alternate methods lets just `git clone` it
+        # The git clone method takes around (1s) depending on repo size.  Not sure if the
+        # other methods take longer or shorter but I preserved the legacy code for now.
```
On higher latency/lower bandwidth connections, a single HTTP call will always be faster than git cloning. They also do not require `git` to be on command line.

logicminds commented:
Regarding finding the git command.  We would need to implement a which command.  Something similar to facter's which command. https://github.com/puppetlabs/facter/blob/main/lib/facter/custom_facts/core/execution.rb#L39

This might entail another PR for implementing a utility class for items like this.   Thoughts?

glennsarti commented:
Can also go low tech and just try and run `git --version`?

I don't _think_ we need a helper class right now but I wouldn't say no.

logicminds commented:
> Can also go low tech and just try and run `git --version`? 
> 
> I don't _think_ we need a helper class right now but I wouldn't say no.

Oh right, the easy way.  Doh!

logicminds commented:
@glennsarti I addressed all the concerns

1. Check for git
2. Add configuration for clone dir, defaults to OS temp path
3. Tests are passing on windows

glennsarti commented:
Whoops... oh well... was meant to merge #34.  Good thing 34 is just a docs change 😇
